conformance/test-server/src/main.rs

```diff
@@ -0,0 +1,256 @@
+        _ => {
+            eprintln!("unknown handler");
+            exit(1);
+        }
+    };
+
+    let mut handles = vec![];
+    if transport == "tcp" || transport == "both" {
+        let (_, tcp_handle) = tcp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(tcp_handle);
+    }
+    if transport == "udp" || transport == "both" {
+        let (_, udp_handle) = udp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(udp_handle);
+    }
+    let _ = future::join_all(handles).await;
+
+    Ok(())
+}
+
+#[derive(Parser)]
+#[clap(name = "Message responder test DNS server")]
+struct Args {
+    #[clap(default_value = "base", long = "handler")]
+    handler: String,
+    #[clap(default_value = "53", long = "port")]
+    port: u16,
+    #[clap(default_value = "both", long = "transport")]
+    transport: String,
+}
+
+async fn udp_server(
+    bind_addr: SocketAddr,
+    handler: HandlerFn,
+) -> std::io::Result<(SocketAddr, JoinHandle<()>)> {
+    let udp = UdpSocket::bind(bind_addr).await?;
+    let peer = udp.local_addr()?;
+    let handle = tokio::task::spawn(async move {
+        loop {
+            let mut buf = [0u8; 4096];
+            let to = match udp.recv_from(&mut buf).await {
+                Ok((_, to)) => to,
+                Err(_) => continue,
+            };
+
+            let _ = match handler(&buf, Transport::Udp) {
+                Ok(Some(resp)) => udp.send_to(&resp[..], to).await,
+                Ok(None) | Err(_) => continue,
+            };
+        }
+    });
+
+    Ok((peer, handle))
+}
+
+async fn tcp_server(
+    bind_addr: SocketAddr,
+    handler: HandlerFn,
+) -> std::io::Result<(SocketAddr, JoinHandle<()>)> {
+    let tcp = TcpListener::bind(bind_addr).await?;
+    let peer = tcp.local_addr()?;
+    let handle = tokio::task::spawn(async move {
+        loop {
+            let (mut stream, _) = tcp.accept().await.unwrap();
+            tokio::task::spawn(async move {
+                loop {
+                    let mut len = [0u8; 2];
+                    match stream.read(&mut len).await {
+                        Ok(..) => {}
+                        Err(e) => {
+                            println!("read error: {e:?}");
+                            break;
+                        }
+                    }
+                    let len = u16::from_be_bytes(len) as usize;
+                    let mut read_buf = vec![0u8; len];
+
+                    if len == 0 {
+                        println!("no message");
+                        break;
+                    }
+
+                    match stream.read_exact(&mut read_buf).await {
+                        Ok(..) => {}
+                        Err(e) => {
+                            println!("read error: {e:?}");
+                            break;
+                        }
+                    }
+
+                    let resp = match handler(&read_buf, Transport::Tcp) {
```

Comment:
I think it would be helpful for debugging to try parsing incoming and outgoing messages and log them (or failing that, the decode error).